### PR TITLE
HPCC-32469 Add support for non-keyed index filtering

### PR DESCRIPTION
**Pull request by jakesmith**
Rowservice index reads with filters on non-keyed fields failed.

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: rtl/eclrtl/rtlnewkey.cpp [hunk 1]
```diff
@@ -2245,6 +2245,45 @@ void RowFilter::extractMemKeyFilter(const RtlRecord & record, const UnsignedArra
     }
 }
 
+void RowFilter::splitIntoKeyFilter(const RtlRecord & record, RowFilter &keyFilter)
+{
+    if (0 == filters.ordinality())
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply rtl/eclrtl/rtlnewkey.cpp:2250]: Could also sort the filters member if that was important, rather than cloning.

**Review by ghalliday — APPROVED** [review]
@jakesmith the code looks fine.  It may all be necessary, but my impression is that it is doing more work than it needs to - particularly the creation of the wildFieldFilters.  Would a IndexRowFilter be better?
Happy to merge as is - let me know if you want to revisit, or just merge as-is.

File: rtl/eclrtl/rtlnewkey.cpp [hunk 2]
```diff
@@ -2249,39 +2249,22 @@ void RowFilter::splitIntoKeyFilter(const RtlRecord & record, RowFilter &keyFilte
...
+        const IFieldFilter & cur = filters.item(i);
+        if (cur.queryFieldIndex() < numKeyedFields)
         {
-            addFilter(OLINK(temp.item(filterIdx)));
-            filterIdx++;
+            keyFilters.append(OLINK(cur));
+            filters.remove(i);
         }
     }
+    keyFilter.sortByFieldOrder(); // NB: need to be ordered ahead of keyFilter.createSegmentMonitors being used
```
Comment (ghalliday) [inline rtl/eclrtl/rtlnewkey.cpp:2264]: line 2260 has added to keyFilters not keyFilter - so the keyed filters will be thrown away

**Review by ghalliday — CHANGES_REQUESTED** [review]

**Comment by AttilaVamos** [comment]
On Rocky-8 environment 2 tests failed:
```
    350. Test: indexlimits.ecl ( version: optRemoteRead=true )
    '--- indexlimits.xml
    +++ indexlimits.xml
    @@ -34,11 +34,8 @@
      <Row><fname>KEYED LIMIT EXC</fname><lname>KEYED LIMIT EXC</lname><age>0</age><addrs></addrs><__internal_fpos__>0</__internal_fpos__></Row>
     </Dataset>
     <Dataset name=indexReadNoHitKeyedLimit>
    - <Row><fname>Charles        </fname><lname>Dance          </lname><age>60</age><addrs><Row><addr>somewhere</addr></Row></addrs><__internal_fpos__>0</__internal_fpos__></Row>
    - <Row><fname>Charles        </fname><lname>Gould          </lname><age>50</age><addrs><Row><addr>somewhere</addr></Row></addrs><__internal_fpos__>0</__internal_fpos__></Row>
     </Dataset>
     <Dataset name=choosen_indexReadNoHitKeyedLimit>
    - <Row><fname>Charles        </fname><lname>Dance          </lname><age>60</age><addrs><Row><addr>somewhere</addr></Row></addrs><__internal_fpos__>0</__internal_fpos__></Row>
     </Dataset>
     <Dataset name=count_indexReadHitKeyedLimit>
      <Row><count_indexReadHitKeyedLimit>0</count_indexReadHitKeyedLimit></Row>
    @@ -47,10 +44,10 @@
      <Row><count_indexReadComboHitRowLimit>0</count_indexReadComboHitRowLimit></Row>
     </Dataset>
     <Dataset name=count_indexReadNoHitKeyedLimit>
    - <Row><count_indexReadNoHitKeyedLimit>2</count_indexReadNoHitKeyedLimit></Row>
    + <Row><count_indexReadNoHitKeyedLimit>0</count_indexReadNoHitKeyedLimit></Row>
     </Dataset>
     <Dataset name=count_choosen_indexReadNoHitKeyedLimit>
    - <Row><count_choosen_indexReadNoHitKeyedLimit>1</count_choosen_indexReadNoHitKeyedLimit></Row>
    + <Row><count_choosen_indexReadNoHitKeyedLimit>0</count_choosen_indexReadNoHitKeyedLimit></Row>
     </Dataset>
     <Dataset name=normIndexReadHitLocalLimit>
     </Dataset>
    @@ -72,9 +69,6 @@
      <Row><addr>KEYED LIMIT EXCEEDED</addr></Row>
     </Dataset>
     <Dataset name=normIndexReadNoHitKeyedLimit>
    - <Row><addr>somewhere</addr></Row>
    - <Row><addr>somewhere</addr></Row>
     </Dataset>
     <Dataset name=choosen_normIndexReadNoHitKeyedLimit>
    - <Row><addr>somewhere</addr></Row>
     </Dataset>
    '
    348. Test: indexlimit3.ecl ( version: optRemoteRead=true )
    '--- indexlimit3.xml
    +++ indexlimit3.xml
    @@ -29,15 +29,13 @@
      <Row><key>1</key><key2>2</key2><v>8</v><kkeys><Row><ckey>994</ckey><info></info></Row></kkeys><info></info><__internal_fpos__>0</__internal_fpos__></Row>
     </Dataset>
     <Dataset name=Result 10>
    - <Row><key>1</key><key2>2</key2><v>6</v><kkeys><Row><ckey>992</ckey><info></info></Row></kkeys><info></info><__internal_fpos__>0</__internal_fpos__></Row>
    - <Row><key>1</key><key2>2</key2><v>7</v><kkeys><Row><ckey>993</ckey><info></info></Row></kkeys><info></info><__internal_fpos__>0</__internal_fpos__></Row>
    - <Row><key>1</key><key2>2</key2><v>8</v><kkeys><Row><ckey>994</ckey><info></info></Row></kkeys><info></info><__internal_fpos__>0</__internal_fpos__></Row>
    + <Row><key>0</key><key2>0</key2><v>0</v><kkeys></kkeys><info>**KEYED LIMIT HIT IN INDEXREAD**</info><__internal_fpos__>0</__internal_fpos__></Row>
     </Dataset>
     <Dataset name=Result 11>
      <Row><Result_11>3</Result_11></Row>
     </Dataset>
     <Dataset name=Result 12>
    - <Row><Result_12>3</Result_12></Row>
    + <Row><Result_12>1</Result_12></Row>
     </Dataset>
     <Dataset name=Result 13>
      <Row><Result_13>true</Result_13></Row>
    '
```

**Review by ghalliday — APPROVED** [review]

**Pull request merged**